apps/scratch/src/pages/Edit.tsx

```diff
@@ -0,0 +1,145 @@
+  AppIFrameMessage,
+  AppIFrameResponse,
+} from "../../../../frontend/src/types/app-iframe-message";
+import { patchScratchVm } from "../vm";
+
+const respondToMessageEvent = (
+  event: MessageEvent,
+  message: Omit<AppIFrameResponse, "id" | "type">,
+) => {
+  if (!event.source) {
+    console.error("Cannot respond to event without source:", event);
+    return;
+  }
+
+  event.source.postMessage(
+    {
+      id: event.data.id,
+      type: "response",
+      ...message,
+    } as AppIFrameMessage,
+    {
+      targetOrigin: event.origin,
+    },
+  );
+};
+
+const Edit = () => {
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const { sessionId, taskId } = useParams();
+  const [hasLoaded, setHasLoaded] = useState(false);
+  const [isInIframe, setIsInIframe] = useState(false);
+  const [vm, setVm] = useState<VM | null>(null);
+
+  useEffect(() => {
+    if (window.parent && window.parent !== window) {
+      setIsInIframe(true);
+    }
+
+    setHasLoaded(true);
+  }, []);
+
+  // a callback function that listens for messages from the parent window
+  const handleParentWindowMessage = useCallback(
+    async (event: MessageEvent) => {
+      if (event.source !== window.parent) {
+        return;
+      }
+
+      const message = event.data as AppIFrameMessage;
+
+      if (message.type === "response") {
+        console.error(
+          `Response in AppIFrameMessage message: '${JSON.stringify(message)}'`,
+        );
+        return;
+      }
+
+      switch (message.procedure) {
+        case "getHeight":
+          respondToMessageEvent(event, {
+            procedure: "getHeight",
+            result: document.body.scrollHeight,
+          });
+          break;
+        case "getSubmission":
+          if (vm) {
+            respondToMessageEvent(event, {
+              procedure: "getSubmission",
+              result: new Blob([vm.toJSON()], {
+                type: "application/json",
+              }),
+            });
+          }
+          break;
+        case "getTask":
+          if (vm) {
+            vm.saveProjectSb3().then((content) => {
+              respondToMessageEvent(event, {
+                procedure: "getTask",
+                result: content,
+              });
+            });
+          }
+          break;
+        case "loadTask":
+          if (vm) {
+            const sb3Project = await message.arguments.arrayBuffer();
+
+            vm.loadProject(sb3Project).then(() => {
+              respondToMessageEvent(event, {
+                procedure: "loadTask",
+              });
+            });
+          }
+          break;
+        default:
+          break;
+      }
+    },
+    [vm],
+  );
+
+  // add an event listener to listen for messages from the iFrame
+  useEffect(() => {
+    if (isInIframe) {
+      window.addEventListener("message", handleParentWindowMessage);
+
+      return () => {
+        window.removeEventListener("message", handleParentWindowMessage);
+      };
```

Comment:
yup, all this logic should really be moved away from the Scratch app itself, even if for now it's just a separate module in this project. It'll make it simpler / more maintainable / more readable if all of this is encapsulated within a small, tidy lib with a clean API.

I'm thinking we might want/need two versions: (1) that has the logic but is entirely React-unaware, (2) a React-aware wrapper for that logic.

Reply:
Yes I fully agree :)